modules/compatcheck/main.go

```
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"context"
+	"dagger/compatcheck/internal/dagger"
+	"fmt"
+	"runtime"
+
+	"github.com/dagger/dagger/dagql/introspection"
+	"github.com/josephburnett/jd/v2"
+	"github.com/moby/buildkit/identity"
+	"github.com/tidwall/gjson"
+	"golang.org/x/exp/rand"
+	"golang.org/x/mod/semver"
+)
+
+type Compatcheck struct{}
+
+// Compare the schema of given module with different versions of engine
+func (m *Compatcheck) Run(ctx context.Context,
+	// ref of the module to compare
+	module,
+	// version of engine to compare with
+	versionA,
+	// version of engine to compare to
+	versionB string,
+	//+optional
+	// only required when one of the version to compare is 'dev' [optional]
+	source *dagger.Directory) error {
+	if (versionA == "dev" || versionB == "dev") && source == nil {
+		return fmt.Errorf("--source flag is required when one of the requested engine version is 'dev'")
+	}
+
+	schemaA, err := m.getSchemaForModuleForEngineVersion(ctx, module, versionA, source)
+	if err != nil {
+		return err
+	}
+
+	schemaB, err := m.getSchemaForModuleForEngineVersion(ctx, module, versionB, source)
+	if err != nil {
+		return err
+	}
+
+	a, _ := jd.ReadJsonString(schemaA)
+	b, _ := jd.ReadJsonString(schemaB)
```

nit: don't discard the errors, return them.